metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/data_resolver.py

```diff
@@ -0,0 +1,660 @@
+from datahub.ingestion.source.powerbi.config import Constant
+from datahub.ingestion.source.powerbi.rest_api_wrapper.data_classes import (
+    Dashboard,
+    Page,
+    PowerBIDataset,
+    Report,
+    Tile,
+    User,
+    Workspace,
+    new_powerbi_dataset,
+)
+
+# Logger instance
+logger = logging.getLogger(__name__)
+
+
+def is_permission_error(e: requests.exceptions.HTTPError) -> bool:
+    return e.response.status_code == 401 or e.response.status_code == 403
+
+
+class DataResolverBase(ABC):
+    SCOPE: str = "https://analysis.windows.net/powerbi/api/.default"
+    BASE_URL: str = "https://api.powerbi.com/v1.0/myorg/groups"
+    ADMIN_BASE_URL: str = "https://api.powerbi.com/v1.0/myorg/admin"
+    AUTHORITY: str = "https://login.microsoftonline.com/"
+    TOP: int = 1000
+
+    def __init__(
+        self,
+        client_id: str,
+        client_secret: str,
+        tenant_id: str,
+    ):
+        self.__access_token: str = ""
+        self.__tenant_id = tenant_id
+        # Power-Bi Auth (Service Principal Auth)
+        self.__msal_client = msal.ConfidentialClientApplication(
+            client_id,
+            client_credential=client_secret,
+            authority=DataResolverBase.AUTHORITY + tenant_id,
+        )
+
+        # Test connection by generating access token
+        logger.info("Trying to connect to {}".format(self._get_authority_url()))
+        self.get_access_token()
+        logger.info("Connected to {}".format(self._get_authority_url()))
+        self._request_session = requests.Session()
+        # set re-try parameter for request_session
+        self._request_session.mount(
+            "https://",
+            HTTPAdapter(
+                max_retries=Retry(
+                    total=3,
+                    backoff_factor=1,
+                    allowed_methods=None,
+                    status_forcelist=[429, 500, 502, 503, 504],
+                )
+            ),
+        )
+
+    @abstractmethod
+    def get_groups_endpoint(self) -> str:
+        pass
+
+    @abstractmethod
+    def get_dashboards_endpoint(self, workspace: Workspace) -> str:
+        pass
+
+    @abstractmethod
+    def get_reports_endpoint(self, workspace: Workspace) -> str:
+        pass
+
+    @abstractmethod
+    def get_tiles_endpoint(self, workspace: Workspace, dashboard_id: str) -> str:
+        pass
+
+    @abstractmethod
+    def _get_pages_by_report(self, workspace: Workspace, report_id: str) -> List[Page]:
+        pass
+
+    @abstractmethod
+    def get_dataset(
+        self, workspace_id: str, dataset_id: str
+    ) -> Optional[PowerBIDataset]:
+        pass
+
+    @abstractmethod
+    def get_users(self, workspace_id: str, entity: str, entity_id: str) -> List[User]:
+        pass
+
+    def _get_authority_url(self):
+        return "{}{}".format(DataResolverBase.AUTHORITY, self.__tenant_id)
+
+    def get_access_token(self):
+        if self.__access_token != "":
```

Comment:
Will this ever be None?

Reply:
lets set to none to minimize the number of states we need to deal with

Reply:
done